### ISSUE TITLE
Using relation to create another relation produces error

### ISSUE THREAD

**Issue by stayallive** [nunomaduro/larastan#2111]
- Larastan Version: 3.0.0
- Laravel Version: 11.32.0

### Description

When returning a relationship that uses another relationship an error is shown.

### Laravel code where the issue was found

```php
<?php

class User extends Model 
{
    /** @return \Illuminate\Database\Eloquent\Relations\BelongsToMany<\App\Models\Team, $this> */
    public function teams(): BelongsToMany
    {
        return $this->belongsToMany(Team::class);
    }
    /** @return \Illuminate\Database\Eloquent\Relations\BelongsToMany<\App\Models\Team, $this> */
    public function paidTeams(): BelongsToMany
    {
        return $this->teams()->whereHas('activeSubscriptions');
    }
}
```

Error:

```
  157    Method App\Models\User::paidTeams() should return                                                                                
         Illuminate\Database\Eloquent\Relations\BelongsToMany<App\Models\Team, $this(App\Models\User)> but returns    
         Illuminate\Database\Eloquent\Relations\BelongsToMany<App\Models\Team,                                                            
         Illuminate\Database\Eloquent\Relations\BelongsToMany<App\Models\Team, $this(App\Models\User)>>.              
         🪪  return.type 
```

Let me know if it's helpful to provide more information but I believe I followed the instruction in the upgrade guide correctly and applied the correct type hints but the error confuses me so it might be me doing something wrong anyway and would be happy to be educated since I was unable to find an answer myself.

**Comment by canvural** [comment]
Hi,

I'll try to take a look, but it's low priority issue I'd say. It's safe to ignore for now.

**Comment by jstoks** [comment]
Adding any additional queries causes this issue. It looks like a recursive type declaration of `$this`. It's unclear if this is actually a bug with Larastan or part of Phpstan itself. 

```php
/** @return HasMany<Note, $this> */
public function notes() {
    return $this->hasMany(Note::class);
}

 /**
 * @return HasMany<Note, $this>
*/
public function firstNote(): HasMany {
    return $this->notes()
        ->withoutGlobalScopes()
        ->select(DB::raw('distinct on(reservation_id) *'))
        ->orderBy('reservation_id', 'asc')
        ->orderBy('created_at', 'asc');
}
```

```
 137    Method App\Models\Reservation::firstNote() should return Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,               
         Illuminate\Database\Eloquent\Model> but returns Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,                        
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note, Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,        
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note, Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,        
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note, Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,        
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note, Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,        
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note, Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,        
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note, Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,        
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note, Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note,        
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Note, $this(App\Models\Reservation)>>>>>>>>>>>>>>>>.                         
         🪪  return.type
```

As a workaround I'm just adding a `@var` declaration to force the return type.

**Comment by stayallive** [comment]
Oh, interesting. For the `@var` I need to put it as a temp variable first right?

This does seem odd, if this is a phpstan bug happy to report it over there but I would not know how to boil this down to the kind of magic docblocks for vanilla phpstan 😄 

I know this isn't critical and can be ignored but this does make the upgrade harder than it has to be because new errors are reported after the upgrade and I need to inspect each and every error to see if I can leave it or ignore it (because it's this issue) which consumes a great deal of time (in addition to all the comments that need to be added, which can be automated but still, I have hundreds of them to check to see if they are right and adjust). 

But I appreciate your quick update @canvural and would love to see it fixed but can also understand this is not a priority ❤️

**Comment by calebdw** [comment]
I encountered and fixed this in #1990, so this is likely due to the partial application of that PR (or a regression during the v3 updates).

**Comment by jstoks** [comment]
> Oh, interesting. For the `@var` I need to put it as a temp variable first right?

I'm not having to do that. I just put it above the return statement.
